Add config file instructions and reorganize a bit

jashapiro opened the pull request:
Closes #112 

This PR adds more detail about config files and profiles to the external user documentation. 

⚠️ As I was writing, I made some organizational changes to that section and the section about AWS, so this PR should be evaluated before we get too far into revisions, so that those organizational changes don't cause too many conflicts! ⚠️
(I retargeted #120 to this branch, as I think it makes sense to finish this one first.)

The organizational changes are mostly about hierarchy: I moved most everything up a level, but nested the "profile" discussion within the more general configuration discussion. It seemed like talking about workflow parameters in a config file was the easiest transition in, so I did that first, then moved on to the discussion of profiles.

I also moved the two required parameters up to the very top of the "Configuring `scpca-nf`" section, as I want to make sure that people don't miss those. (I had some thoughts about how to make it more clear that those are required at the command line, but I think that will be part of #105, and I will put those thoughts there.)

I didn't include too much detail about profiles as it seems like there were a number of ways one could go on that front, and most would not apply to most people. If you think there are places that I should expand a bit more though, I would be happy to do so!

allyhawkins commented on external-data-instructions.md:
```diff
@@ -2,70 +2,105 @@
-### Prepare the metadata file 
+This will pull the `scpca-nf` workflow directly from Github, using the `v0.2.4` version, and run it based on the settings in the local configuration file `my_config.config`.
+
+**Note:** `spca-nf` is under active development.
```
```suggestion
**Note:** `scpca-nf` is under active development.
```
Minor typo

allyhawkins commented on external-data-instructions.md:
```diff
@@ -2,70 +2,105 @@
+Local running may be sufficient for small jobs or testing, but you will most likely want to run your workflow in a high performance computing environment (HPC), such as an institutional computing cluster or on a cloud service like AWS.
+To do this, we recommend using [Nextflow profiles](https://www.nextflow.io/docs/latest/config.html#config-profiles) to encapsulate settings like the [`executor`](https://www.nextflow.io/docs/latest/executor.html) that will be used to run each process and associated details that may be required, such as queue names or the container engine your system uses. 
+
+In our example template file [user_template.config](https://github.com/AlexsLemonade/scpca-nf/blob/main/examples/user_template.config), we define a profile named `cluster` which could be invoked with the following command:
```
I think it would be worth adding a sentence about using singularity here. Our setup assumes usage of docker, so I think it would be useful to state that if your system uses singularity and does not use docker, then that will need to be reflected in the config file and maybe even show an example or link to the [nextflow docs on using singularity](https://www.nextflow.io/docs/latest/singularity.html). I think this will be a common enough use that it would be helpful information to have.

jashapiro replied:
I hinted at this in the sentence before "associated details... such as container engine" but I will expand that a bit, and also add Singularity to the prereq options.

I didn't want to specifically call out too much about the example file here, to preserve our ability to modify it if we wanted later. That was also part of the logic behind changing the name to the generic "cluster". My thought was that people are very much likely to need to read the full documentation, so I don't want to provide the illusion of completeness, that they might be able to just take the version we provide and have it "just work".

allyhawkins replied:
I am good with the change to `cluster` and understand about not wanting to use the example too much. I think linking to the nextflow docs on singularity would be helpful though or at least a statement saying that our default container settings are to use docker so if you are using something else, like singularity, that needs to be reflected in your config file.

allyhawkins commented on external-data-instructions.md:
```diff
@@ -2,70 +2,105 @@
+
+### Configuration files
+
+Workflow parameters can also be set in a configuration file by setting the values `params.run_metafile` and `params.outdir` as follows: 
```
I keep going back and forth on this, but I wonder if it would be helpful to have an intro into what a configuration file is and also state that they require the `.config` extension. I feel like right now we just jump into adding parameters to the config file without a lot of explanation, but then again maybe it doesn't need explanation and the links you provided in the later section are enough.

jashapiro replied:
I added a link to the documetation for the config file here, but the minimal version really is just a list of parameters, so I wanted to just show that example first. I wasn't clear if the way I did it here was the best; I can try to play with wording. 

A note though... the file doesn't need to have a `.config` extension; that is only if is going to be found automatically as a `nextflow.config` file. I do want every example to have that extension, as I think it is a good idea, but it isn't a requirement.

allyhawkins reviewed (commented):
I think this looks pretty good overall. I'm okay with the changes to the organization and they make sense. My biggest comment is about adding in some information about using singularity, since I think that will be fairly common for people using an HPC and is also present in the config file. 

I had one other comment about adding an intro sentence before you jump into adding parameters to the config files, but I could go either way on that one.

jashapiro commented on external-data-instructions.md:
````diff
@@ -2,70 +2,105 @@
+
+```groovy
+// my_config.config
+params.run_metafile = '<path/to/metadata_file>'
````
@allyhawkins I used the `params.variable` convention here, because it seemed more straightforward in this simple example. But I was wondering if you thought we should perhaps use that also for `user_template.config` for consistency? I feel like the blocks/bracket syntax is good when things get more complex, but it may be simpler to reserve that for the profile section, and have the "universal" settings with the simpler (but more redundant?) syntax.

allyhawkins replied:
I was thinking about this and wondering if it was going to be confusing to have two separate version of it, but agree that here it makes sense to use `params.variable`. I am good with the change to the template to be inline with using `params.variable` syntax and be consistent then.

jashapiro commented:
@allyhawkins I think I addrssed most of your comments, and now link to the Singularity instructions in two places. (I also removed the direct docker link, instead linking to the nextflow instructions there as well, for consistency. I was originally going to link directly to singularity, but their docs are terrible for initial users.) 

I also updated the template file a bit, trying to format it to make the organization a bit more clear, and removing the `params{}` block in favor of `params.variable`. 

I hope this addresses the suggestions you had made, but let me know if I missed anything!

allyhawkins commented on examples/user_template.config:
```diff
@@ -1,37 +1,44 @@
+/*
+Global parameters for scpca-nf
+------------------------------
+The values presented here are examples, and should be replaced in with 
+the correct paths to any indicated files or other values that need to be set.
```
```suggestion
The values presented here are examples, and should be replaced with 
the correct paths to any indicated files or other values that need to be set.
```
Minor typo.

allyhawkins commented on examples/user_template.config:
```diff
@@ -1,37 +1,44 @@
+Here we provide an example using slurm as the executor system. 
+For information about available process executors and their options, see https://www.nextflow.io/docs/latest/executor.html
+This profile should be adjusted based on the details of your system and can be invoked at runtime with 
+`--profile cluster` (or whatever name you have chosen)
```
```suggestion
`-profile cluster` (or whatever name you have chosen)
```
I think this should be a single dash.

allyhawkins reviewed (approved):
These changes look good to me! I just picked up on two minor typos that need to be fixed and then its good to go.
